### PR TITLE
Conductor and sequencer p2p refactoring

### LINKED ISSUE #10936: op-node: Improve shutdown behavior
* Add (more) info-logging on shutdown
* Revisit order in which components are shut down
    *  e.g. the we may want to stop sequencing before stopping p2p distribution of blocks
* Fix https://github.com/ethereum-optimism/optimism/issues/8086

### PR DESCRIPTION
**Pull request by anacrolix**
Addresses https://github.com/ethereum-optimism/optimism/issues/10936#issuecomment-2176330084.

* Stops the sequencer before stopping p2p so it can communicate any last work.
* Reworks the p2p enabled logic so it's behaving like p2p is disabled when it's only in a partially shutdown state.
* Increases logging around sequencing management in conductor.
* Tidies up context use in various places to avoid the conductor sleeping during shutdown or getting stuck in API calls.
* Removes a check for host that's no longer optional.

### CODE REVIEW

**Comment by tynes** [comment]
very nice

File: op-conductor/conductor/service.go
Comment (anacrolix) [inline op-conductor/conductor/service.go:651]: After an error, the service would wait to retry before actually attending to a shutdown. It loops back and then immediately shuts down anyway.

File: op-conductor/conductor/service.go
Comment (anacrolix) [inline op-conductor/conductor/service.go:704]: I think there's a race here to handle shutdown or stop the sequencer. Previously it could be asked to shutdown then it would try to stop the sequencer indefinitely.

File: op-conductor/conductor/service.go
Comment (anacrolix) [inline op-conductor/conductor/service.go:707]: I've added extra logging here because I want to know what the conductor sees after it stops the sequencer.

File: op-e2e/setup.go
Comment (anacrolix) [inline op-e2e/setup.go:418]: This is a new error case that pops up due to the node actually stopping when asked.

File: op-conductor/conductor/service.go
Comment (sebastianst) [inline op-conductor/conductor/service.go:652]: Wouldn't you now wait two times for a `oc.retryBackoff()`? You call a `time.Sleep` when entering this block. So maybe we want to do this instead?
```suggestion
		default:
```

File: op-conductor/conductor/service.go
Comment (sebastianst) [inline op-conductor/conductor/service.go]: I'm not sure, but maybe the reason `strings.Contains` was used instead of `errors.Is` is because the error is reconstructed from a json rpc response, so it wouldn't have the actual error type. This should be double checked. If it is indeed of the correct error type, then your fix is better!

File: op-conductor/conductor/service.go
Comment (sebastianst) [inline op-conductor/conductor/service.go]: Good to loop in an op-conductor expert to answer this question. My gut feeling is that it's more important to shut down a sequencer rather than shutting down op-conductor quickly, since it's always better to have 0 instead of >1 sequencers 😅

File: op-node/node/node.go
Comment (sebastianst) [inline op-node/node/node.go:473]: You removed a `|| p2pNode == nil` branch in the `if` check, is it guaranteed that `p2p.NewNodeP2P` returns a non-nil instance if err==nil?

File: op-node/node/node.go
Comment (sebastianst) [inline op-node/node/node.go:46]: Why is this actually better? It seems more robust to me to check at runtime whether a field is nil instead of assuming that the `OpNode` has been created with the correct constructor and configuration? Couldn't there be scenarios in e.g. testing where an OpNode struct is directly created and then it's safer to just test directly e.g. `n.p2pNode != nil`?

So in that sense maybe the node `Config` is something that is consumed by the constructor but not necessarily valid or required after.

File: op-node/node/node.go
Comment (sebastianst) [inline op-node/node/node.go]: Don't you also need to check that it's actually enabled?
```suggestion
	return n.cfg.P2P != nil && !n.cfg.P2P.Disabled()
```

File: op-node/node/node.go
Comment (sebastianst) [inline op-node/node/node.go]: To answer your question, yes I think it is guaranteed because `initHeartbeat` is called after `initP2P`, which sets up  p2p if it's enabled.

I think this check here is actually missing a `nil` check for the `cfg.P2P` field. If you retain the config in the `OpNode` struct and add ` && !n.cfg.P2P.Disabled()` to `OpNode.p2pEnabled()`, then you could use `p2pEnabled()` instead. Or use `cfg.P2PEnabled()` like I suggest in another comment.

File: op-node/node/node.go
Comment (sebastianst) [inline op-node/node/node.go]: It may be nicer to add this method directly to the `Config` struct as a public method `P2PEnabled() bool`. Then you could use it in setup code instead and don't need to retain the `cfg` in the `OpNode` struct.

File: op-node/node/node.go
Comment (sebastianst) [inline op-node/node/node.go]: You could do this, and then don't need to retain the config in the OpNode struct:
```suggestion
	if n.l2Driver != nil {
		latestHead, err := n.l2Driver.StopSequencer(ctx)
		switch {
		case errors.Is(err, driver.ErrSequencerNotEnabled): // do nothing, also not log
```
`ErrSequencerNotEnabled` is returned by the `DisabledSequencer` on `Stop`.

File: op-conductor/conductor/service.go
Comment (anacrolix) [reply op-conductor/conductor/service.go:652]: Ah, no I've just forgotten to remove the actual time.Sleep. Thanks

File: op-conductor/conductor/service.go
Comment (anacrolix) [reply op-conductor/conductor/service.go]: Yes, knowledge sharing ftw. Who's the knowledge expert for op-conductor?

File: op-conductor/conductor/service.go
Comment (anacrolix) [reply op-conductor/conductor/service.go]: You're right. I was staring at StopSequencer in op-node that I forgot about the RPC. I'll revert this part.

File: op-node/node/node.go
Comment (anacrolix) [reply op-node/node/node.go:46]: The reason is that checking for nil is defensive programming. We have a clear predicate for when p2p is enabled so we should use that. To phrase it another way, nil-checking/defensive programming creates additional program states that we can't and don't account for. p2pNode != nil is being used here for testing if it's been shutdown but also for whether p2p is expected to exist and that partial state propagates to other areas of the code that are trying to call in with events.

File: op-node/node/node.go
Comment (anacrolix) [reply op-node/node/node.go]: Thanks, I looked into it a bit further and tightened it up a bit. Looks like the p2p nil check was missed.

File: op-node/node/node.go
Comment (anacrolix) [reply op-node/node/node.go]: I think you're right, thanks.

Comment (anacrolix) [reply op-node/node/node.go]: I see why I did it, it's because the same reason I wanted to test the config for features rather than the node state might apply to other features later.

File: op-node/node/node.go
Comment (anacrolix) [reply op-node/node/node.go:473]: I did a deep dive, looks like there's some incomplete refactoring logic here.

File: op-node/node/node.go
Comment (anacrolix) [reply op-node/node/node.go:46]: As it happens, it's always created with the config. Per the other comment about weird states, I think it's worth keeping it in.

File: op-node/node/node.go
Comment (anacrolix) [reply op-node/node/node.go]: It is preferable it fits the convention here. I've fixed it.

File: op-node/node/node.go
Comment (anacrolix) [reply op-node/node/node.go:473]: Ok I've incorporated what I found.

File: op-conductor/conductor/service.go
Comment (sebastianst) [reply op-conductor/conductor/service.go]: @zhwrd and @0x00101010, they're also set as [codeowners](https://github.com/ethereum-optimism/optimism/blob/develop/.github/CODEOWNERS)

File: op-conductor/conductor/service.go
Comment (sebastianst) [reply op-conductor/conductor/service.go]: I don't see it reverted.

File: op-node/node/node.go
Comment (sebastianst) [reply op-node/node/node.go:46]: OK I don't oppose it.

File: op-node/p2p/node.go
Comment (anacrolix) [inline op-node/p2p/node.go:113]: The host check is moved out, I completed the reliance on SetupP2P.Disabled elsewhere as it looks like was intended.

File: op-conductor/conductor/service.go
Comment (zhwrd) [reply op-conductor/conductor/service.go]: So stopSequencer is called after conductor loses leadership. In the event that the stopSequencer call fails, it actually has little real consequences because the sequencer cant produce a block and gossip / commit it to the raft log (requires leadership). Once conductor comes back up it will check its leader and sequencer state and attempt to stop the sequencer again.

So it is "okay" to fail to stop a sequencer, the state will eventually be rectified and we wont have two active sequencers that are actually producing blocks.

Comment (anacrolix) [reply op-conductor/conductor/service.go]: Thanks! In that case it makes sense to pass a regular context in here and prevent the conductor getting stuck on an API call (like for instance during shutdown).

Comment (anacrolix) [reply op-conductor/conductor/service.go]: I've incorporated your very useful comment into the PR.

File: op-node/p2p/node.go
Comment (anacrolix) [reply op-node/p2p/node.go:183]: This code isn't modified from the original, so I'll ignore this for now.

**Comment by tynes** [comment]
I am not sure how extensive the test coverage of this code is. Do we have unit tests? Is the best way to test just syncing a node?

**Review by zhwrd — APPROVED** [review]

**Comment by anacrolix** [comment]
I'll check tests and fix conflicts now.

**Comment by anacrolix** [comment]
I don't think the test failures are related to this code

**Comment by anacrolix** [comment]
Looks like I need a code owners approval?

File: op-node/node/node.go
Comment (sebastianst) [inline op-node/node/node.go]: I still think that it makes sense to make this a method on the `Config` instead. You could then call it with `n.cfg.P2PEnabled()`.

File: op-node/p2p/node.go
Comment (sebastianst) [inline op-node/p2p/node.go:1]: [Semantic diff](https://app.semanticdiff.com/gh/ethereum-optimism/optimism/pull/11455/changes#op-node/p2p/node.go) was extremely helpful here to confirm that the only change that happens inside `init` was to remove the `if n.host != nil {` branch conditional.

**Review by sebastianst — COMMENTED** [review]
LGTM, effectively the only change in behavior in this PR is to call `n.l2Driver.StopSequencer(ctx)` inside `OpNode.Stop`. Everything else is just clean up work.

Will approve once the open conversations are resolved.

File: op-node/p2p/node.go
Comment (anacrolix) [reply op-node/p2p/node.go:1]: That's a great website

File: op-node/node/config.go
Comment (sebastianst) [inline op-node/node/config.go]:
```suggestion
func (cfg *Config) P2PEnabled() bool {
```

File: op-node/node/node.go
Comment (sebastianst) [inline op-node/node/node.go]:
```suggestion
	return n.cfg.P2PEnabled()
```

**Review by sebastianst — APPROVED** [review]

**Comment by anacrolix** [comment]
🎉

**Pull request merged**
